### PR TITLE
Backport pagetitleprovider and seo

### PR DESCRIPTION
**Pull request by DanielSiepmann**

### CODE REVIEW

File: Classes/Frontend/MetaInformation/DateMetaInformationService.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    public function __construct(
+        MetaTagManagerRegistry $metaTagManagerRegistry,
+        DateTitleProviderInterface $titleProvider
+    ) {
+        $this->metaTagManagerRegistry = $metaTagManagerRegistry;
+        $this->titleProvider = $titleProvider;
+    }
+
+    public function setDate(Date $date): void
```
Comment (extcode) [inline Classes/Frontend/MetaInformation/DateMetaInformationService.php:55]: I don't think the `setDate()` method is a good choice from the outside. When the IDE suggests the method to me, I don't know what it really does. I don't realize that this method overwrites the `description` and the `keywords`. Wouldn't `initialize()` be a better option for this method?

File: Classes/Frontend/MetaInformation/EventMetaInformationService.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    public function __construct(
+        MetaTagManagerRegistry $metaTagManagerRegistry,
+        EventTitleProviderInterface $titleProvider
+    ) {
+        $this->metaTagManagerRegistry = $metaTagManagerRegistry;
+        $this->titleProvider = $titleProvider;
+    }
+
+    public function setEvent(Event $event): void
```
Comment (extcode) [inline Classes/Frontend/MetaInformation/EventMetaInformationService.php:55]: I don't think the `setEvent()` method is a good choice from the outside. When the IDE suggests the method to me, I don't know what it really does. I don't realize that this method overwrites the `description` and the `keywords`. Wouldn't `initialize()` be a better option for this method?

**Comment by DanielSiepmann** [comment]
Do not squash, rebase instead.
This is only a backport of an existing feature from v4.

File: Classes/Frontend/MetaInformation/DateMetaInformationService.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    private function setDescription(Date $date): void
+    {
+        $description = '';
+        if ($date->getEvent() !== null) {
+            $description = $date->getEvent()->getTeaser();
+        }
+        if ($description === '') {
+            return;
+        }
```
Comment (extcode) [inline Classes/Frontend/MetaInformation/DateMetaInformationService.php:71]: I find this easier to read:
```suggestion
        if ($date->getEvent() === null) {
            return;
        }
        $description = $date->getEvent()->getTeaser();
        if ($description === '') {
            return;
        }
```
but that is a matter of taste. In later PHP versions you could certainly write this even more simply with `??`.

File: Classes/Frontend/PageTitleProvider/DateTitleProvider.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    public function getTitle(): string
+    {
+        if ($this->date === null) {
+            return '';
+        }
+
+        return implode(' ', [
+            $this->date->getEvent()->getTitle(),
+            $this->date->getStart()->format('d.m.Y H:i'),
```
Comment (extcode) [inline Classes/Frontend/PageTitleProvider/DateTitleProvider.php:48]: This is a typical German format for the output of dates. Can this simply be made configurable, or does the extension use this date format so hard in various places?

**Review by extcode — APPROVED** [review]

**Pull request merged**